src/librustc_mir/build/scope.rs

```diff
@@ -123,10 +126,17 @@ impl<'a,'tcx> Builder<'a,'tcx> {
             extent: extent.clone(),
             continue_block: loop_block,
             break_block: break_block,
+            might_terminate: false
         };
         self.loop_scopes.push(loop_scope);
         let r = f(self);
-        assert!(self.loop_scopes.pop().unwrap().extent == extent);
+        let loop_scope = self.loop_scopes.pop().unwrap();
+        assert!(loop_scope.extent == extent);
+        // Once we’re done with the loop, we want to check if we have to “assign” the loop’s
+        // destination lvalue.
+        if loop_scope.might_terminate {
+            self.cfg.push_assign_unit(break_block, span, destination);
+        }
```

Comment:
Hmm, I did not expect *this* code to push statements onto the basic block. It violates my mental model, where basic blocks are either "things to append instructions onto" or "things to branch to", but not both. Given that this function is called exactly once, and with `R=()`, maybe we should just change the return type to `BlockAnd<bool>` and document that it returns true if we may break out of this loop. Then we can move this "push_assign_unit" up into the caller (the arm for `ExprKind::Loop`).